### PR TITLE
[Issue #707] implement stream storage

### PR DESCRIPTION
**Pull request by huasiy**
part of  #707

### CODE REVIEW

File: pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package io.pixelsdb.pixels.common.physical;
```
Comment (bianhq) [inline pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java:1]: Add license header.

File: pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public class StreamPath
+{
+    private String host;
+    private int port;
+    public boolean valid = false;
+
+    public StreamPath(String path)
+    {
+        requireNonNull(path);
+        if (path.contains(":///"))
```
Comment (bianhq) [inline pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java:14]: why ':///' instread of '://'?

File: pixels-common/src/main/java/io/pixelsdb/pixels/common/utils/HttpServerHandler.java [hunk 3]
```diff
@@ -100,4 +101,8 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
         ChannelFuture f = ctx.writeAndFlush(response);
         f.addListener(ChannelFutureListener.CLOSE);
     }
+
+    public void setServerCloser(Runnable serverCloser) {
```
Comment (bianhq) [inline pixels-common/src/main/java/io/pixelsdb/pixels/common/utils/HttpServerHandler.java:105]: New line after method.

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
+package io.pixelsdb.pixels.storage.stream;
```
Comment (bianhq) [inline pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java:1]: License header.

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    @Override
+    public boolean supportsAsync() { return false; }
+
+    @Override
+    public CompletableFuture<ByteBuffer> readAsync(long offset, int len) throws IOException
+    {
+        throw new UnsupportedOperationException("Can't get file length in PhysicalStreamReader");
```
Comment (bianhq) [inline pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java:99]: Can't get file length?

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public long getFileLength() throws IOException
+    {
+        throw new UnsupportedOperationException("Can't get file length in PhysicalStreamReader");
+    }
+
+    @Override
+    public void seek(long desired) throws IOException
+    {
+        throw new UnsupportedOperationException("Can't get file length in PhysicalStreamReader");
```
Comment (bianhq) [inline pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/PhysicalStreamReader.java:44]: Can't get file length?

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamInputStream.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
+package io.pixelsdb.pixels.storage.stream.io;
```
Comment (bianhq) [inline pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamInputStream.java:1]: License header.

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
+package io.pixelsdb.pixels.storage.stream.io;
```
Comment (bianhq) [inline pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamOutputStream.java:1]: License header.

**Review by bianhq — CHANGES_REQUESTED** [review]

File: pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java (hunk 2 above)
Comment (huasiy) [reply pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java:14]: refer to FilePath.java
```java
public FilePath(String path)
    {
        requireNonNull(path);
        if (path.startsWith("file:///"))
        {
            valid = true;
            realPath = path.substring(path.indexOf("://") + 3);
        }
        else if (path.startsWith("/"))
        {
            valid = true;
            realPath = path;
        }

        if (valid)
        {
            File file = new File(realPath);
            isDir = file.isDirectory();
        }
    }
```

Comment (bianhq) [reply pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java:14]: Local file paths always start with '/,' hence there are three /// in total.
However, streaming paths should start with the destination hostname.

File: pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamInputStream.java (hunk 7 above)
Comment (bianhq) [reply pixels-storage/pixels-storage-stream/src/main/java/io/pixelsdb/pixels/storage/stream/io/StreamInputStream.java:1]: The license header is still not added.

File: pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java (hunk 2 above)
Comment (bianhq) [reply pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/StreamPath.java:14]: Local file paths always start with '/,' hence there are three /// in total.
However, streaming paths should start with the destination hostname.

**Review by bianhq — CHANGES_REQUESTED** [review]

**Review by bianhq — APPROVED** [review]
LGTM.

**Pull request merged**